colossalai/zero/sharded_model/sharded_grad.py

```diff
@@ -0,0 +1,122 @@
+import os
+from typing import List, Optional, Tuple
+from xml.dom.expatbuilder import parseFragment
+
+import torch
+import torch.distributed as dist
+from colossalai.context.parallel_mode import ParallelMode
+from colossalai.core import global_context as gpc
+from torch.distributed import ProcessGroup
+from torch.nn.parameter import Parameter
+
+from ._zero3_utils import (alloc_storage, assert_in_engine, free_storage,
+                           get_shard)
+
+
+class ShardedGradient:
+    def __init__(self,
+                 param: Parameter,
+                 process_group: Optional[ProcessGroup] = None,
+                 offload_config: Optional[dict] = None
+                 ) -> None:
+        self.param = param
+        self.process_group = process_group or gpc.get_group(ParallelMode.DATA)
+        self.shard_idx = self.process_group.rank()
+        self.num_shards = self.process_group.size()
+        self.offload_config = offload_config
+
+        self._cpu_offload = offload_config.get('device', None) == 'cpu' if offload_config else False
+
+        sharded_param, _ = get_shard(param.data, self.shard_idx, self.num_shards)
```

Comment:
SharedParam is another Class. You do not need do it here?

Reply:
Also you shall not assume the parameter.data has been gathered! Therefore, use its size is not appropriate.